modules/aggregation-layers/src/aggregation-layer-v9/aggregation-layer.ts

```diff
@@ -0,0 +1,111 @@
+  static layerName = 'AggregationLayer';
+
+  state!: {
+    gpuAggregator: GPUAggregator | null;
+    cpuAggregator: CPUAggregator | null;
+  };
+
+  /** Allow this layer to participates in the draw cycle */
+  get isDrawable() {
+    return true;
+  }
+
+  /** Called to create a GPUAggregator instance */
+  abstract getGPUAggregator(): GPUAggregator | null;
+  /** Called to create a CPUAggregator instance if getGPUAggregator() returns null */
+  abstract getCPUAggregator(): CPUAggregator | null;
+  /** Called when some attributes change, a chance to mark Aggregator as dirty */
+  abstract onAttributeChange(id: string): void;
+
+  initializeState(): void {
+    this.getAttributeManager()!.remove(['instancePickingColors']);
+  }
+
+  // Override Layer.updateState to update the GPUAggregator instance
+  updateState(params: UpdateParameters<this>) {
+    super.updateState(params);
+
+    if (params.changeFlags.extensionsChanged) {
+      this.state.gpuAggregator?.destroy();
+      this.state.gpuAggregator = this.getGPUAggregator();
+      if (this.state.gpuAggregator) {
+        this.getAttributeManager()!.invalidateAll();
+      } else if (!this.state.cpuAggregator) {
+        this.state.cpuAggregator = this.getCPUAggregator();
+      }
+    }
+  }
+
+  // Override Layer.finalizeState to dispose the GPUAggregator instance
+  finalizeState(context: LayerContext) {
+    super.finalizeState(context);
+    this.state.gpuAggregator?.destroy();
+  }
+
+  // Override Layer.updateAttributes to update the aggregator
+  protected updateAttributes(changedAttributes: {[id: string]: Attribute}) {
+    this.getAggregator()?.setProps({
+      attributes: changedAttributes
+    });
+
+    for (const id in changedAttributes) {
+      this.onAttributeChange(id);
+    }
+  }
+
+  draw({moduleParameters}) {
+    // GPU aggregation needs `moduleSettings` for projection/filter uniforms which are only accessible at draw time
+    // GPUAggregator's Buffers are allocated during `updateState`/`GPUAggregator.setProps`
+    // and passed down to the sublayer attributes in renderLayers()
+    // Although the Buffers have been bound to the sublayer's Model, their content are not populated yet
+    // GPUAggregator.update() is called in the draw cycle here right before Buffers are used by sublayer.draw()
+    this.state.gpuAggregator?.update({moduleSettings: moduleParameters});
+  }
+
+  protected getAggregator(): Aggregator | null {
+    return this.state.gpuAggregator || this.state.cpuAggregator;
+  }
+
+  // override CompositeLayer._getAttributeManager to create AttributeManager instance
+  _getAttributeManager() {
+    return new AttributeManager(this.context.device, {
+      id: this.props.id,
+      stats: this.context.stats
+    });
+  }
+
+  // Override CompositeLayer._postUpdate to update attributes and the CPUAggregator
+  protected _postUpdate(updateParams: UpdateParameters<this>, forceUpdate: boolean) {
```

Comment:
It's not "encouraged" for subclasses to override these underscore methods. The underscore kind of signifies "don't touch me unless you know what you are doing".

`_postUpdate` specifically is defined in three classes:
- Layer: update attributes; update models with attribute changes; set model instance count; update picking module props from auto-highlighting props
- CompositeLayer: render sublayers
- AggregationLayer: update attributes and render sublayers

The code that goes in this method is guaranteed to run AFTER subclass-defined `updateState`.

If we keep finding new places where we need to override it, maybe it will be a good idea to add an official new lifecycle method, but then we'll need to define it better e.g. what does "post" mean, and what goes into `postUpdateState` but not `updateState`?